### PR TITLE
Reprocessing detection feature flag

### LINKED ISSUE #5759: Control inconsistency detection with an experimental feature flag
**Is your feature request related to a problem? Please describe.**

Inconsistency detection during reprocessing is an important feature to guarantee upgrades can be safely done, as well as to guarantee the health of your Zeebe cluster. However, it is currently buggy (see https://github.com/zeebe-io/zeebe/issues/5688 for some info). It is still a very useful feature, but for now we should enable it purely when verifying upgrades such that we don't affect the stability of a normal run.

**Describe the solution you'd like**

I'd like to control whether or not inconsistency detection is enabled, which allows users to make use of it where it's most needed - during an upgrade. I would propose to make this an experimental flag to give us some flexibility on the future of the upgrade process, which still needs to be simplified and streamlined to guarantee a good user experience.

**Describe alternatives you've considered**

Fixing all bugs until the detection is correct - a laudable goal which we should achieve, but time consuming at the moment. Adding a flag will buy us some well needed time to achieve it.

**Additional context**

Affected versions would be 0.24.4 and 0.25.0 - we should release this feature as the only fix in the upcoming patch releases, 0.24.5 and 0.25.1, and users should upgrade as soon as possible.

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description

Introduces a new feature flag in the experimental config to enable the detection of reprocessing inconsistency. Per default it is now disabled.

## Related issues

closes https://github.com/zeebe-io/zeebe/issues/5759

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: dist/src/main/config/broker.standalone.yaml.template [hunk 1]
```diff
@@ -524,3 +524,8 @@
       # Sets the maximum batch size, which is send per append request to a follower.
       # This setting can also be overridden using the environment variable ZEEBE_EXPERIMENTAL_MAX_APPEND_BATCH_SIZE
       # maxAppendBatchSize = 32KB;
+
+      # Enables the detection of an inconsistency during reprocessing. If a inconsistency is detect the StreamProcessor is
+      # failed and the partition becomes unhealthy, now further progress will made on that specific partition.
```
Comment (npepinpe) [inline dist/src/main/config/broker.standalone.yaml.template:529]:
```suggestion
      # failed and the partition becomes unhealthy, no further progress will made on that specific partition.
```

File: dist/src/main/config/broker.standalone.yaml.template [hunk 2]
```diff
@@ -524,3 +524,8 @@
...
+      # This setting can also be overridden using the environment variable ZEEBE_EXPERIMENTAL_DETECT_REPROCESSING_INCONSISTENCY
+      # detectReprocessingInconsistency = 32KB;
```
Comment (npepinpe) [inline dist/src/main/config/broker.standalone.yaml.template:531]:
```suggestion
      # detectReprocessingInconsistency = false;
```

**Review by npepinpe — COMMENTED** [review]
just quickly noticed two typos

File: engine/src/test/java/io/zeebe/engine/util/EngineRule.java [hunk 3]
```diff
@@ -346,6 +367,19 @@ private void onNewEventCommitted() {
...
 
+  private final class ReprocessingCompletedListener implements StreamProcessorLifecycleAware {
+    private final ActorFuture<Void> reprocessingComplete = new CompletableActorFuture<>();
+
+    @Override
+    public void onRecovered(final ReadonlyProcessingContext context) {
+      reprocessingComplete.complete(null);
+    }
+
+    public void awaitReprocessingComplete() {
```
Comment (npepinpe) [inline engine/src/test/java/io/zeebe/engine/util/EngineRule.java:378]: Can we add a timeout? Otherwise we could end up with a test that never finishes

**Review by npepinpe — COMMENTED** [review]
Thanks for the quick work, looks great :tada: 

One small concern I have is with the test you added - should the conditions you create there not lead to inconsistencies, then the test would pass even if the flag was ignored. I see two ways - either way find a way to assert that the reprocessing is simply not called (e.g. mock, ideally something less hacky :sweat_smile:), or we start once with and once without, asserting that it does detect the inconsistency only when the flag is on. I imagine you also thought about it, so maybe you can ease my concern with your arguments :slightly_smiling_face:

**Comment by Zelldon** [comment]
> One small concern I have is with the test you added - should the conditions you create there not lead to inconsistencies, then the test would pass even if the flag was ignored. I see two ways - either way find a way to assert that the reprocessing is simply not called (e.g. mock, ideally something less hacky sweat_smile), or we start once with and once without, asserting that it does detect the inconsistency only when the flag is on. I imagine you also thought about it, so maybe you can ease my concern with your arguments slightly_smiling_face

Lets talk about it tomorrow. I currently don't get it. The feature flag completely turns off detection, so I will not detect it per default.

**Review by npepinpe — APPROVED** [review]
As discussed, let's keep it as is for now - if we do a second pass, we can extract the detection logic to its own component, which will not only make that easier to test, but also make the test for the flag much more robust by decoupling it from the detection logic.

**Pull request merged**